reana_workflow_controller/rest/workflows.py

```diff
@@ -888,3 +891,258 @@ def get_workflow_retention_rules(workflow_id_or_name: str, user: str):
+              message:
+                type: string
+              errors:
+                  type: array
+                  items:
+                    type: string
+          examples:
+            application/json:
+              {
+                "message": "Internal controller error.",
+                "errors": ["Internal controller error."]
+              }
+    """
+    try:
+        user = User.query.filter(User.id_ == user_id).first()
+        if not user:
+            return (
+                jsonify({"message": f"User with id '{user_id}' does not exist."}),
+                404,
+            )
+
+        if (
+            re.match(
+                r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,7}\b",
+                user_email_to_share_with,
+            )
+            is None
+        ):
+            raise ValueError(f"User email '{user_email_to_share_with}' is not valid.")
+
+        if user.email == user_email_to_share_with:
+            raise ValueError("Unable to share a workflow with yourself.")
+
+        user_to_share_with = (
+            Session.query(User).filter(User.email == user_email_to_share_with).first()
+        )
+
+        if not user_to_share_with:
+            return (
+                jsonify(
+                    {
+                        "message": f"User with email '{user_email_to_share_with}' does not exist."
+                    }
+                ),
+                404,
+            )
+
+        if valid_until:
+            try:
+                datetime.date.fromisoformat(valid_until)
+            except ValueError as e:
+                raise ValueError(
+                    f"Date format is not valid ({str(e)}). Please use YYYY-MM-DD format."
+                )
+
+            # check if date is in the future
+            if datetime.date.fromisoformat(valid_until) < datetime.date.today():
+                raise ValueError("The 'valid_until' date cannot be in the past.")
+
+        if message and len(message) > 5000:
+            raise ValueError(
+                "Message is too long. Please keep it under 5000 characters."
+            )
+
+        workflow = _get_workflow_with_uuid_or_name(workflow_id_or_name, str(user_id))
+
+        existing_share = (
+            Session.query(UserWorkflow)
+            .filter_by(user_id=user_to_share_with.id_, workflow_id=workflow.id_)
+            .first()
+        )
+
+        if existing_share:
+            return (
+                jsonify(
+                    {
+                        "message": f"{workflow.get_full_workflow_name()} is already shared with {user_email_to_share_with}."
+                    }
+                ),
+                409,
+            )
+
+        Session.add(
+            UserWorkflow(
+                user_id=user_to_share_with.id_,
+                workflow_id=workflow.id_,
+                message=message,
+                valid_until=valid_until,
+            )
+        )
+        Session.commit()
```

Comment:
There is a potential race condition here, as this can happen:
1. `existing_share` is fetched from the database, but it does not exist so `None` is returned
2. another (concurrent) API request creates the workflow share
3. we try to create a new share, but `Session.add`/`Session.commit` will fail as the share already exists

Note that in this case nothing bad happens, as the database will simply return an integrity error (given that there will be multiple rows with the same primary keys). However, we could prevent this issue by simply inserting the new row without checking first if it exists:
```python
try:
    Session.add(UserWorkflow(...))
    Session.commit()
except IntegrityError:
    Session.rollback()
    return jsonify(...), 409
```

Also see https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use

Reply:
Nicely spotted! :+1: 

The change has been applied to this superseding PR: #552